### ISSUE TITLE
🚀 Feature: Weaviate Instrumentation

### ISSUE THREAD

**Issue by nirga** [traceloop/openllmetry#243]
### Which component is this feature for?

All Packages

### 🔖 Feature description

Instrument calls to [Weaviate](https://weaviate.io/developers/weaviate/client-libraries/python_v3), including adding attributes, similarly to our [Chroma](https://github.com/traceloop/openllmetry/tree/main/packages/opentelemetry-instrumentation-chroma) instrumentation. The instrumentation should support all types of calls - streaming, non streaming, async, etc.

### 🎤 Why is this feature needed?

Completeness of OpenLLMetry

### ✌️ How do you aim to achieve this?

Similar to other instrumentations, we have in this repo.

### 🔄️ Additional Information

* Make sure to initialize this in [Traceloop SDK](https://github.com/traceloop/openllmetry/blob/main/packages/traceloop-sdk/traceloop/sdk/tracing/tracing.py#L241).
* Add proper [tests](https://github.com/traceloop/openllmetry/tree/main/packages/traceloop-sdk/tests).

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### Are you willing to submit PR?

None

**Comment by paolorechia** [comment]
I’ll look into this if I get some time this week 😊

**Comment by mohitk1995** [comment]
It would be great if we have the above support.

**Comment by paolorechia** [comment]
Hi, I started working on this, but I've ran into some issues with dependency conflicts. Basically, haystack depends on `pydantic < 2`, and `weaviate-client>4` depends on `pydantic>=2`.

While haystack v2 (beta) seems to have dropped pydantic as a dependency (https://github.com/deepset-ai/haystack/blob/main/pyproject.toml), we have the choice to instrument instead weaviate-client v3 instead (https://weaviate.io/developers/weaviate/client-libraries/python_v3).

The downside is that weaviate-client v3 API is not as neatly organized as the v4. And also means some duplicated effort, as I started building the sample app with the v4 interface.

On the other hand, upgrading haystack to v2 is risky, as it's beta. And would require much more rework, since we would have to update the instrumentation.

So weaviate-client v3 seems like the best idea - unless you have a better idea.
